File: src/main/java/com/natu/ftax/transaction/presentation/EditFieldRequest.java
package com.natu.ftax.transaction.presentation;

import com.fasterxml.jackson.annotation.JsonFormat;
import com.natu.ftax.transaction.domain.EditTransactionCommand;
import jakarta.validation.constraints.NotNull;
import lombok.Getter;

import java.math.BigDecimal;
import java.time.Instant;

@Getter
public class EditFieldRequest {


    @NotNull(message = "Id cannot be null")
    private String id;

    private String transactionType;

    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd'T'HH:mm:ss.SSSXXX")
    private Instant instant;

    private String tokenIn;
    private String tokenOut;
    private String tokenFee;

    private BigDecimal amountIn;
    private BigDecimal amountOut;
    private BigDecimal amountFee;

    private String externalId;

    public EditTransactionCommand toCommand() {
        return new EditTransactionCommand(
            id,
            transactionType,
            instant,
            tokenIn,
            tokenOut,
            tokenFee,
            amountIn,
            amountOut,
            amountFee,
            externalId
        );
    }

    // Getters and setters
    // ...
}

File: src/main/kotlin/com/natu/ftax/transaction/presentation/EditFieldRequest.kt
package com.natu.ftax.transaction.presentation

import com.fasterxml.jackson.annotation.JsonFormat
import com.fasterxml.jackson.annotation.JsonProperty
import com.natu.ftax.transaction.domain.EditTransactionCommand
import com.natu.ftax.transaction.domain.TransactionType
import jakarta.validation.constraints.PastOrPresent
import jakarta.validation.constraints.PositiveOrZero
import org.jetbrains.annotations.NotNull
import java.util.*


class EditFieldRequest(

    @field:NotNull(value = "Id cannot be null")
    val id: String,

    @JsonProperty("transactionType")
    val transactionType: TransactionType?,

    @field:PastOrPresent(message = "Date must be in the past or present")
    @field:JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd'T'HH:mm:ss.SSSXXX")
    val date: Date?,

    val tokenIn: String?,
    val tokenOut: String?,
    val tokenFee: String?,

    @field:PositiveOrZero(message = "Amount must be positive")
    val amountIn: Double?,

    @field:PositiveOrZero(message = "Amount must be positive")
    val amountOut: Double?,

    @field:PositiveOrZero(message = "Fee must be positive")
    val amountFee: Double?,

    val externalId: String?


) {
    fun toCommand() = EditTransactionCommand(
        id = id,
        transactionType = transactionType,
        date = date,
        tokenIn = tokenIn,
        tokenOut = tokenOut,
        tokenFee = tokenFee,
        amountIn = amountIn,
        amountOut = amountOut,
        amountFee = amountFee,
        externalId = externalId
    )
}
